palace/linalg/errorestimator.cpp

```diff
@@ -0,0 +1,395 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "errorestimator.hpp"
+#include "fem/coefficient.hpp"
+#include "fem/integrator.hpp"
+#include "fem/multigrid.hpp"
+#include "models/materialoperator.hpp"
+#include "utils/communication.hpp"
+#include "utils/errorindicators.hpp"
+#include "utils/iodata.hpp"
+
+namespace palace
+{
+
+using namespace fem;
+
+CurlFluxErrorEstimator::CurlFluxErrorEstimator(
+    const IoData &iodata, const MaterialOperator &mat_op,
+    const std::vector<std::unique_ptr<mfem::ParMesh>> &mesh,
+    mfem::ParFiniteElementSpace &fes)
+  : mat_op(mat_op), fes(fes),
+    smooth_flux_fecs(ConstructFECollections<mfem::ND_FECollection>(
+        iodata.solver.order, mesh.back()->Dimension(), iodata.solver.linear.mg_max_levels,
+        iodata.solver.linear.mg_coarsen_type, iodata.solver.linear.pc_mat_lor)),
+    smooth_flux_fes(ConstructFiniteElementSpaceHierarchy<mfem::ND_FECollection>(
+        iodata.solver.linear.mg_max_levels, iodata.solver.linear.mg_legacy_transfer,
+        iodata.solver.pa_order_threshold, mesh, smooth_flux_fecs)),
+    smooth_projector(smooth_flux_fes, iodata.solver.linear.tol, 200, 0,
+                     iodata.solver.pa_order_threshold),
+    coarse_flux_fec(iodata.solver.order, mesh.back()->Dimension(),
+                    mfem::BasisType::GaussLobatto),
+    coarse_flux_fes(mesh.back().get(), &coarse_flux_fec, mesh.back()->Dimension()),
+    scalar_mass_matrices(fes.GetNE()), smooth_to_coarse_embed(fes.GetNE())
+{
+  mfem::MassIntegrator mass_integrator;
+  for (int e = 0; e < fes.GetNE(); e++)
+  {
+    // Loop over each element, and save an elemental mass matrix.
+    // Will exploit the fact that vector L2 mass matrix components are independent.
+    const auto &coarse_fe = *coarse_flux_fes.GetFE(e);
+    auto &T = *coarse_flux_fes.GetElementTransformation(e);
+    mass_integrator.AssembleElementMatrix(coarse_fe, T, scalar_mass_matrices[e]);
+
+    const auto &smooth_fe = *smooth_flux_fes.GetFinestFESpace().GetFE(e);
+    coarse_fe.Project(smooth_fe, T, smooth_to_coarse_embed[e]);
+  }
+}
+
+IndicatorsAndNormalization CurlFluxErrorEstimator::operator()(const ComplexVector &v,
```

Comment:
Seems like this is a generalization of the real-valued case. Can this be templated and the functions combined? This should be possible, see for example most of the solvers in `linalg/`.